src/utils/util/resolver-pipeline/oscal-profile-resolve-merge.xsl

```diff
@@ -77,83 +90,93 @@
     </xsl:template>
 
-    <xsl:key name="control-by-id" match="control" use="@id"/>
-
-    <xsl:template match="node() | @*" mode="o:custom-merge">
-        <xsl:copy>
-            <xsl:apply-templates mode="#current" select="node() | @*"/>
-        </xsl:copy>
-    </xsl:template>
-
+    <!-- Process children but do not copy <custom> tags themselves. -->
     <xsl:template match="custom" mode="o:custom-merge">
         <xsl:apply-templates mode="#current"/>
     </xsl:template>
 
+    <!-- Apply the requested ordering.
+        Creating the properly combined sequence of inserted controls
+        is in downstream template. -->
     <xsl:template match="insert-controls" mode="o:custom-merge">
-        <xsl:variable name="inserted-controls">
+        <xsl:variable name="inserted-controls" as="element()*">
             <xsl:apply-templates mode="#current"/>
         </xsl:variable>
         <xsl:variable name="keep-order" select="not(@order = ('descending','ascending'))" as="xs:boolean"/>
         <xsl:variable name="sort-order" select="@order[.='descending'],'ascending'"/>
         <!-- Setting sort-key to '1' sorts into given order -->
-        <xsl:perform-sort  select="$inserted-controls">
-                <xsl:sort select="if ($keep-order) then '1' else @control-id"/>
+        <xsl:perform-sort select="$inserted-controls">
+            <xsl:sort select="if ($keep-order) then '1' else @id" 
+                order="{$sort-order[1]}"/>
         </xsl:perform-sort>
     </xsl:template>
 
-    <xsl:template match="include-controls" mode="o:custom-merge">
-        <xsl:variable name="match-patterns" select="matching/@pattern"/>
-        <xsl:call-template name="combine-elements">
-            <xsl:with-param name="who" select="key('control-by-id', with-id),
-                /*/selection//control[some $p in ($match-patterns) satisfies (matches(@id,o:glob-as-regex(string($p))))]"/>
-        </xsl:call-template>
-    </xsl:template>
+    <xsl:include href="select-or-custom-merge.xsl"/>
 
-    <xsl:template match="include-all" mode="o:custom-merge">
-        <xsl:variable name="match-patterns" select="matching/@pattern"/>
+    <!-- Combine selected controls.
+        Creating the sequence of controls ($who) is in code shared with
+        selection phase. -->
+    <xsl:template match="include-all | include-controls" mode="o:custom-merge">
+        <xsl:variable name="who" as="element(o:control)*">
+            <xsl:apply-templates select="ancestor::*[last()]//selection" mode="o:select">
+                <xsl:with-param tunnel="yes" name="import-instruction" select="ancestor::insert-controls"/>
+            </xsl:apply-templates>
+        </xsl:variable>
         <xsl:call-template name="combine-elements">
-            <xsl:with-param name="who" select="//control"/>
+            <xsl:with-param name="who" select="$who"/>
         </xsl:call-template>
     </xsl:template>
 
-    <xsl:include href="oscal-profile-resolve-functions.xsl"/>
```

Comment:
Included from select-or-custom-merge.xsl, so it doesn't need to be included from here, too.